From 3072526ae9cedbbfd577af92d3d47350e55f60cd Mon Sep 17 00:00:00 2001
From: Michael Martis <martis@chromium.org>
Date: Mon, 7 Jun 2021 16:06:46 +1000
Subject: [PATCH 4/7] Systemlib rules currently can't handle the systemlib
 build file appearing in the system link dictionary.

This occurs in the gRPC target here:
https://github.com/tensorflow/tensorflow/blob/a4dfb8d1a71385bd6d122e4f27f86dcebb96712d/tensorflow/workspace2.bzl#L639

Gentoo (the primary client of systemlib) only packages TF 2.4.0,
potentially explaining why the new systemlib logic hasn't been exercised
for gRPC.

This change fixes the issue with the systemlib logic.
---
 third_party/repo.bzl | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/third_party/repo.bzl b/third_party/repo.bzl
index 38a8805f31c..344dd745941 100644
--- a/third_party/repo.bzl
+++ b/third_party/repo.bzl
@@ -28,10 +28,11 @@ def _use_system_lib(ctx, name):
     return name in [n.strip() for n in syslibenv.split(",")]
 
 def _get_link_dict(ctx, link_files, build_file):
+    link_dict = {ctx.path(v): ctx.path(Label(k)) for k, v in link_files.items()}
     if build_file:
         # Use BUILD.bazel because it takes precedence over BUILD.
-        link_files = dict(link_files, **{build_file: "BUILD.bazel"})
-    return {ctx.path(v): Label(k) for k, v in link_files.items()}
+        link_dict[ctx.path("BUILD.bazel")] = ctx.path(Label(build_file))
+    return link_dict
 
 def _tf_http_archive_impl(ctx):
     # Construct all labels early on to prevent rule restart. We want the
-- 
2.31.1

